### PR TITLE
Collections remaining pieces

### PR DESCRIPTION
**Pull request by jyeshe**
### Description

This PR adds:
- `:delete_all` route to delete all collection items or to delete matching keys
- `created_after` and `created_before` filters for `:stream_all` and `:stream_match` routes
-  `limit` query parameter applicable also `:stream_all` and `:stream_match` routes
-  an API authentication Plug with the generic part of the JWT verification

Additionally it fixes the query and tests related to the cursor based on `inserted_at` once all streamed items are sorted by this field.
 
### Validation steps

0. `curl --X POST -H "authorization: Bearer ${token}" -H "content-type: application/json" localhost:4000/collections/openhie -d '{"items": [{"key": "key06Nov1", "value": "value1"}, {"key": "key06Nov2", "value": "value2"}]}'
1. `curl -H "authorization: Bearer ${token}" localhost:4000/collections/openhie/key06Nov1` get the created timestamp and use to test stream all and stream match queries with `created_before` and/or `created_after`

### Additional notes for the reviewer

- The reduce to send HTTP chunks was refactored to make it clear the structure of the accumulator (acc) that includes now the custom limit and the `cursor_data` to avoid relying on cursor_count concept.
- begin and finish chunking are now symetrical
- Now that a custom limit will be accepted, ideally the code to create a stream resource should be used to avoid holding the database transaction for too long (after this the database streaming won't be needed anymore) 
    
## AI Usage

Please disclose how you've used AI in this work (it's cool, we just want to know!):

- [x] Code generation (copilot but not intellisense)
- [ ] Learning or fact checking
- [ ] Strategy / design
- [ ] Optimisation / refactoring
- [ ] Translation / spellchecking / doc gen
- [ ] Other
- [ ] I have not used AI

You can read more details in our [Responsible AI Policy](https://www.openfn.org/ai#pull-request-templates)

### Pre-submission checklist

- [x] I have performed a **self-review** of my code.
- [x] I have implemented and tested all related **authorization policies**. (e.g., `:owner`, `:admin`, `:editor`, `:viewer`)
- [ ] I have updated the **changelog**.
- [x] I have ticked a box in "AI usage" in this PR

### CODE REVIEW

File: lib/lightning_web/controllers/collections_controller.ex [hunk 1]
```diff
@@ -118,118 +81,212 @@ defmodule LightningWeb.CollectionsController do
...
-         conn <- begin_chunking(conn) do
+         :ok <- authorize(conn, collection) do
+      key_param = params["key"]
+
+      case Collections.delete_all(collection, key_param) do
+        {:ok, n} ->
+          json(conn, %{key: key_param, deleted: n, error: nil})
+
+        {:error, :not_found} ->
+          json(conn, %{key: key_param, deleted: 0, error: "Items Not Found"})
```
Comment (stuartc) [inline lib/lightning_web/controllers/collections_controller.ex:102]: Going off of the comment above, we don't need to have a split here, no items affected isn't an error, can just be `deleted: n, error: nil`.

File: test/support/factories.ex [hunk 2]
```diff
@@ -329,7 +329,12 @@ defmodule Lightning.Factories do
     %Lightning.Collections.Item{
       collection: build(:collection),
       key: sequence(:key, &"key-#{&1}"),
-      value: sequence(:value, &"value-#{&1}")
+      value: sequence(:value, &"value-#{&1}"),
+      inserted_at:
+        sequence(
+          :inserted_at,
+          &DateTime.add(DateTime.utc_now(), &1, :microsecond)
+        )
```
Comment (stuartc) [inline test/support/factories.ex:337]: ❤️

File: lib/lightning/collections.ex [hunk 3]
```diff
@@ -209,16 +208,57 @@ defmodule Lightning.Collections do
...
+
+  defp filter_by_created_after(query, %{created_after: created_after}),
+    do: where(query, [i], i.inserted_at >= ^created_after)
+
+  defp filter_by_created_after(query, _params), do: query
+
+  defp filter_by_created_before(query, %{created_before: created_before}),
+    do: where(query, [i], i.inserted_at < ^created_before)
+
+  defp filter_by_created_before(query, _params), do: query
```
Comment (stuartc) [inline lib/lightning/collections.ex:256]: Would this benefit from using `Ecto.Query.dynamic/2`?

**Review by stuartc — CHANGES_REQUESTED** [review]
Looking great! Left a few comments, and I see the type check is failing and one test as well.

File: lib/lightning/collections.ex (hunk 3 above)
Comment (jyeshe) [reply lib/lightning/collections.ex:256]: For this single known condition I've followed the approach to keep as simple as possible while making it modular and maintainable. But if we gonna have more complex queries in the future we could make a dynamic reduce in advance.

File: lib/lightning/collections.ex [hunk 4]
```diff
@@ -209,16 +208,57 @@ defmodule Lightning.Collections do
...
+        case key_pattern do
+          nil -> query
+          pattern -> where(query, [i], like(i.key, ^format_pattern(pattern)))
+        end
+      end)
+
+    case Repo.delete_all(query) do
+      {0, nil} -> {:error, :not_found}
+      {n, nil} -> {:ok, n}
+    end
```
[the remark this answers is not in the document]
Comment (jyeshe) [reply lib/lightning/collections.ex]: Done

File: lib/lightning_web/controllers/collections_controller.ex (hunk 1 above)
Comment (jyeshe) [reply lib/lightning_web/controllers/collections_controller.ex]: Done

**Pull request merged**